evalml/pipelines/binary_classification_pipeline.py

```diff
@@ -60,3 +69,34 @@ def _score(X, y, predictions, objective):
         if predictions.ndim > 1:
             predictions = predictions.iloc[:, 1]
         return ClassificationPipeline._score(X, y, predictions, objective)
+
+    def _compute_predictions(self, X, y, objectives, time_series=False):
+        """Compute predictions/probabilities based on objectives."""
+        y_predicted = None
+        y_predicted_proba = None
+        if any(o.score_needs_proba for o in objectives) or (any(not o.score_needs_proba for o in objectives) and self.threshold is not None):
+            y_predicted_proba = self.predict_proba(X, y) if time_series else self.predict_proba(X)
+        if any(not o.score_needs_proba for o in objectives) and self.threshold is None:
+            y_predicted = self._predict(X, y, pad=True) if time_series else self._predict(X)
+        return y_predicted, y_predicted_proba
+
+    def _score_all_objectives(self, X, y, y_pred, y_pred_proba, objectives):
+        scored_successfully = OrderedDict()
+        exceptions = OrderedDict()
+        for objective in objectives:
+            try:
+                if not objective.is_defined_for_problem_type(self.problem_type):
+                    raise ValueError(f'Invalid objective {objective.name} specified for problem type {self.problem_type}')
+                y_pred_to_use = y_pred
+                if self.threshold is not None and not objective.score_needs_proba:
+                    y_pred_to_use = self._predict_with_objective(X, y_pred_proba, objective)
+                score = self._score(X, y, y_pred_proba if objective.score_needs_proba else y_pred_to_use, objective)
+                scored_successfully.update({objective.name: score})
+            except Exception as e:
+                tb = traceback.format_tb(sys.exc_info()[2])
+                exceptions[objective.name] = (e, tb)
+        if exceptions:
```

Comment:
Hmmm, the naming of the variables  and the try/except kind of implied to me that you were going to be forgiving of one or any of the objectives failing.  But I guess what you're trying to do is just get a catalog of what worked and what didn't without raising at the *first* instance of a failed objective.  Am I following that?  So either you get the scores only if everything scores correctly or you get and error detailing what failed and what was successful.

Reply:
Correct! This is actually mostly copied from `PipelineBase` and our current behavior, with lines 91-93 being the difference. But yes, you're spot on. We don't want to error out on the first objective scoring, and want to calculate what we can. We decided this because for AutoML, maybe some objectives just happen to error out but you don't really care about them. In that case, it doesn't make sense to halt search just for this objective, so best to just call that error out and then move on. (When *all* objectives fail though in AutoML though, we scream)

Reply:
I think it'd be best if there was only one `_score_all_objectives` implementation because I want the `PipelineScoreError` logic to only live in one place if we decide to change it in the future.

This works and I don't want to hold up the merge longer but I think we should at least file an issue for refactoring in the future. What do you think @angela97lin @chukarsten ?

Reply:
Did some more refactoring magic, so this is all in one place again 🎈